src/decoder/flac.rs

```diff
@@ -69,63 +309,289 @@ where
             )
             .expect("flac should never have zero channels"),
             total_duration,
+            total_samples,
+            samples_read: 0,
+            is_seekable: settings.is_seekable,
         })
     }
 
+    /// Consumes the decoder and returns the underlying data stream.
+    ///
+    /// This can be useful for recovering the original data source after decoding is complete or
+    /// when the decoder needs to be replaced. The stream position will be at the current
+    /// playback position.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
+    /// use std::fs::File;
+    /// use rodio::decoder::flac::FlacDecoder;
+    ///
+    /// let file = File::open("audio.flac").unwrap();
+    /// let decoder = FlacDecoder::new(file).unwrap();
+    /// let recovered_file = decoder.into_inner();
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if called during a seeking operation when the reader is temporarily `None`.
+    /// This should never happen during normal usage.
     #[inline]
     pub fn into_inner(self) -> R {
-        self.reader.into_inner()
+        self.reader
+            .expect("reader should always be Some")
+            .into_inner()
     }
 }
 
 impl<R> Source for FlacDecoder<R>
 where
     R: Read + Seek,
 {
+    /// Returns the number of samples before parameters change.
+    ///
+    /// For FLAC, this always returns `None` because audio parameters (sample rate, channels, bit
+    /// depth) never change during the stream. This allows Rodio to optimize by not frequently
+    /// checking for parameter changes.
+    ///
+    /// # Implementation Note
+    ///
+    /// FLAC streams have fixed parameters throughout their duration, unlike some formats
+    /// that may have parameter changes at specific points. This enables optimizations
+    /// in the audio pipeline by avoiding frequent parameter validation.
     #[inline]
     fn current_span_len(&self) -> Option<usize> {
         None
     }
 
+    /// Returns the number of audio channels.
+    ///
+    /// FLAC supports 1 to 8 channels. Common configurations:
+    /// - 1 channel: Mono
+    /// - 2 channels: Stereo
+    /// - 6 channels: 5.1 surround
+    /// - 8 channels: 7.1 surround
+    ///
+    /// # Guarantees
+    ///
+    /// The returned value is constant for the lifetime of the decoder and matches
+    /// the channel count specified in the FLAC stream metadata.
     #[inline]
     fn channels(&self) -> ChannelCount {
         self.channels
     }
 
+    /// Returns the sample rate in Hz.
+    ///
+    /// Common rates that FLAC supports are:
```

Comment:
Doesn't flac support every sample range between 1 and ....big number?

In that case this should be: common sample rates